tokio/src/task/builder.rs

```diff
@@ -0,0 +1,95 @@
+#![allow(unreachable_pub)]
+use crate::task::JoinHandle;
+use std::future::Future;
+
+/// Factory which is used to configure the properties of a new task.
+///
+/// Methods can be chained in order to configure it.
+///
+/// Currently, there is only one configuration option:
+///
+/// - [`name`], which specifies an associated name for
+///   the task
+///
+/// There are three types of task that can be spawned from a Builder:
+/// - [`spawn_local`] for executing futures on the current thread
+/// - [`spawn`] for executing [`Send`] futures on the runtime
+/// - [`spawn_blocking`] for executing blocking code in the
+///   blocking thread pool.
+#[derive(Default, Debug)]
+pub struct Builder<'a> {
+    name: Option<&'a str>,
+}
+
+impl Builder<'_> {
+    /// Creates a new task builder
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    /// Assigns a name to the task which will be spawned
+    pub fn name<'b>(&self, name: &'b str) -> Builder<'b> {
+        Builder { name: Some(name) }
+    }
+
+    /// Spawns a task on the executor.
+    ///
+    /// See
+    /// [`task::spawn`][crate::task::spawn] for more details
+    #[cfg_attr(tokio_track_caller, track_caller)]
+    pub fn spawn<T>(self, future: T) -> JoinHandle<T::Output>
```

Comment:
currently, the builder can _only_ be used to spawn tasks on the current runtime. this means that if i want to e.g. spawn tasks on a runtime from outside the runtime context, using a `Handle`, i can't give names to those tasks. it would be nice to have a way to do that.

a potential API for this could look like this:
```rust
let handle: Arc<tokio::runtime::Handle> = /* get a handle from somewhere... */;
let future = async { /* ... */ };

task::Builder::new()
   .named("my-cool-task")
   .spawn_with(future, &handle); // name TBD of course
```
where `Builder::spawn_with` is like `Builder::spawn`, but spawning a task on the provided runtime, rather than the current runtime context.

another option is to add an `Option<&'a Handle>` field to `Builder`, allowing an API like this:
```rust
let handle: Arc<tokio::runtime::Handle> = /* ... */;
let future = async { /* ... */ };

task::Builder::new()
   .named("my-cool-task")
   .runtime(&handle)
   .spawn(future);
```

this has a couple advantages:
  - we wouldn't have to add separate `spawn_with` and `spawn_blocking_with` methods, so the API is a little simpler
  - if a `Builder` is reused for spawning multiple tasks, the handle doesn't need to be passed around to where every task is spawned
  - conceptually, it _feels_ a little better to just set everything as a builder parameter rather than as a new argument

but, the disadvantage is that there is a slight performance hit in the case where we are spawning on the current default runtime, because we'd have to check if the `Option<&Handle>` was set first. this is probably not significant enough to matter, but it's worth thinking about.

similarly, we may also want to think about something similar for `LocalSet`, so that named tasks can be spawned on a `LocalSet` while the `LocalSet` isn't running (using `LocalSet`'s [`spawn_local` method](https://docs.rs/tokio/1.7.1/tokio/task/struct.LocalSet.html#method.spawn_local)).

these APIs would be fine to add in a follow-up PR, IMO, but it's worth thinking about them.

Reply:
The second option seems better to me, and might even be inverted (as an additional interface), like `runtime::build_task().name("name").spawn(async {()})` where that returns a task::Builder with the runtime already set

Reply:
I also noticed that you used `named` instead of name — I kept wanting to use `with_name` or `named` for fluent readability, but I tried to hew closely to [`std::thread::Builder`](https://doc.rust-lang.org/stable/std/thread/struct.Builder.html)'s interface.

Reply:
> I also noticed that you used `named` instead of name — I kept wanting to use `with_name` or `named` for fluent readability, but I tried to hew closely to [`std::thread::Builder`](https://doc.rust-lang.org/stable/std/thread/struct.Builder.html)'s interface, which uses `name` for the chainable/builder/"factory" interface

oh, that wasn't really even intentional --- i think consistency with `thread::Builder` is probably a good choice :)